Remove support for 3DNow!, both intrinsics and builtins.

jyknight opened the pull request:
This set of instructions was only supported by AMD chips starting in
the K6-2 (introduced 1998), and before the "Bulldozer" family
(2011). They were never much used, as they were effectively superseded
by the more-widely-implemented SSE (first implemented on the AMD side
in Athlon XP in 2001).

This is being done as a predecessor towards general removal of MMX
register usage. Since there is almost no usage of the 3DNow!
intrinsics, and no modern hardware even implements them, simple
removal seems like the best option.

(Clang half originally uploaded in https://reviews.llvm.org/D94213)

Works towards issue #41665 and issue #98272.

jyknight committed:
Clang: Remove support for 3DNow!, both intrinsics and builtins.

This set of instructions was only supported by AMD chips starting in
the K6-2 (introduced 1998), and before the "Bulldozer" family
(2011). They were never much used, as they were effectively superseded
by the more-widely-implemented SSE (first implemented on the AMD side
in Athlon XP in 2001).

This is being done as a predecessor towards general removal of MMX
register usage. Since there is almost no usage of the 3DNow!
intrinsics, and no modern hardware even implements them, simple
removal seems like the best option.

Support for the underlying LLVM intrinsics remains, for the
moment. They will be removed in a future patch.

(Originally uploaded in https://reviews.llvm.org/D94213)

Works towards issue #41665.

jyknight committed:
LLVM: Remove support for 3DNow! intrinsics.

This set of instructions was only supported by AMD chips starting in
the K6-2 (introduced 1998), and before the "Bulldozer" family
(2011). They were never much used, as they were effectively superseded
by the more-widely-implemented SSE (first implemented on the AMD side
in Athlon XP in 2001).

This is being done as a predecessor towards general removal of MMX
register usage. Since there is almost no usage of the 3DNow!
intrinsics, and no modern hardware even implements them, simple
removal seems like the best option.

Works towards issue #41665.

phoebewang reviewed (commented):
This is much like what we have removed for KNL intrinsics/instructions., so general LGTM. But I'd like Simon to sign off given I'm not familar with 3DNOW instructions.

RKSimon commented:
I'd prefer we rip off ALL the 3dnow/mmx bandaid in one big series of patches - and not split across the 19.X/20.X releases. 

So can we get all the mmx patches done in time for the 19.0 branch?

Also, this "might" have bigger impact than the KNL changes - should we announce the plan in 19.X and do it for 20.X?

RKSimon commented on clang/lib/Headers/mm3dnow.h:
```diff
@@ -7,151 +7,16 @@
 
+// 3dNow intrinsics are no longer supported, and this header remains only as a
+// stub for users who were including it to get to _m_prefetch or
+// _m_prefetchw. Such uses should prefer x86intrin.h.
```
How best can we add a deprecation warning to builds for people who are doing this?

jyknight replied:
Looking at this again, I do think it's reasonable to emit a `#warning` here.

I was initially worried that might cause issues for users, but there's so very very few users of anything 3dnow-related to start with, and the ones I found generally conditioned the include on `__3dNOW__` being defined, anyhow.

RKSimon commented on llvm/lib/Target/X86/X86.td:
```diff
@@ -1825,32 +1818,32 @@ def : ProcModel<P, SapphireRapidsModel,
 def : Proc<"k6",   [FeatureX87, FeatureCX8, FeatureMMX],
                    [TuningSlowUAMem16, TuningInsertVZEROUPPER]>;
-def : Proc<"k6-2", [FeatureX87, FeatureCX8, Feature3DNow],
+def : Proc<"k6-2", [FeatureX87, FeatureCX8],
```
These still need FeatureMMX and SSEPrefetch (or a similar flag)

jyknight replied:
Oops! I had fixed that but in a not-yet-committed diff in my checkout. (Tests fail, otherwise!)

jyknight committed:
Fix bugs:
- Add FeatureMMX where needed.
- Add mayLoad, mayStore, hasSideEffects as appropriate, where they were previously inferred from the pattern.

jyknight committed:
Fix clang release notes formatting.

RKSimon commented on llvm/lib/Target/X86/X86.td:
```diff
@@ -1825,32 +1818,32 @@ def : ProcModel<P, SapphireRapidsModel,
 def : Proc<"k6",   [FeatureX87, FeatureCX8, FeatureMMX],
                    [TuningSlowUAMem16, TuningInsertVZEROUPPER]>;
-def : Proc<"k6-2", [FeatureX87, FeatureCX8, Feature3DNow],
+def : Proc<"k6-2", [FeatureX87, FeatureCX8, FeatureMMX],
```
I think the best we can do is add FeaturePRFCHW as well?

jyknight replied:
Done.

jyknight committed:
Mark mm3dnow.h header as deprecated.

jyknight committed:
Also add the prfchw feature where needed.

jyknight commented:
> I'd prefer we rip off ALL the 3dnow/mmx bandaid in one big series of patches - and not split across the 19.X/20.X releases.

The 3dnow part is a lot easier -- because it's not trying to be compatible, just dropping support entirely. And I don't think anyone will care, since 3dnow is effectively never used (even before it was discontinued in 2010, it was barely used).

For that same reason -- that it's unused and nobody will notice -- I also think there's not really a reason to wait on removing 3dnow support until other MMX changes are done.

> So can we get all the mmx patches done in time for the 19.0 branch?

Maybe? Depends on how the discussion on the other PR goes -- whether that proposal is acceptable, or if a different transition strategy is desired.

topperc commented:
Was `3dnow` never part of `sys:getHostCPUFeatures` for -march=native? I expected you would need to remove it from there, but it looks like it isn't there

topperc commented:
This LGTM. The prefetch mess was one thing I knew was a little complicated, but it looks like you've taken care of that.

topperc commented on llvm/test/CodeGen/X86/prefetch.ll:
```diff
@@ -6,16 +6,11 @@
-; 3dnow prefetch instruction will only get used if you have no other prefetch instructions enabled
-
-; rdar://10538297
+; supporting prefetchw implicitly provides prefetcht0/1/2/nta as well, as we need something to fall back to for the non-write hint.
```
Do we have a test for this comment?

topperc replied:
nevermind that's the `-mattr=+prfchw` RUN line. It was collapsed in the github view so I didn't see it.

RKSimon commented:
> > So can we get all the mmx patches done in time for the 19.0 branch?
> 
> Maybe? Depends on how the discussion on the other PR goes -- whether that proposal is acceptable, or if a different transition strategy is desired.

Apart from #96540 how many more patches do you envision (still need to remove all the llvm side of things, plus the actual MMX builtins in BuiltinsX86.def - what else?). Should we just put all of that in #96540?

jyknight commented:
I've written down an overall plan into issue #98272. I'm not sure if everything can be finished by the LLVM 19 branch date, but I don't think there's any reason to wait on this CL. I'd like to just go ahead with it.

phoebewang reviewed (approved):
LGTM.

RKSimon commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -936,6 +936,24 @@ X86 Support
+    ``_m_pfnacc``, ``_m_pfpnacc``, ``_m_pi2fw``, ``_m_pswapdsf``,
+    ``_m_pswapdsi``.
+  * The compiler builtins corresponding to each of the above
+    intrinsics have also been removed  (``__builtin_ia32_femms``, and so on).
```
Please can you make it clear that inline-asm is still supported

RKSimon commented on llvm/lib/Target/X86/X86Schedule.td:
```diff
@@ -481,7 +481,7 @@ defm WriteAESKeyGen : X86SchedWritePair<ReadAfterVecXLd>; // Key Generation.
 // Carry-less multiplication instructions.
 defm WriteCLMul : X86SchedWritePair<ReadAfterVecXLd>;
 
-// EMMS/FEMMS
```
keep these - FEMMS still uses this sched class

RKSimon commented on llvm/lib/Target/X86/X86ScheduleZnver3.td:
```diff
@@ -1301,7 +1301,7 @@ defm : Zn3WriteResXMMPair<WriteAESKeyGen, [Zn3FPAES01], 4, [1], 1>; // Key Gener
 // Carry-less multiplication instructions.
 defm : Zn3WriteResXMMPair<WriteCLMul, [Zn3FPCLM01], 4, [4], 4>;
 
-// EMMS/FEMMS
```
keep these - FEMMS still uses this sched class

RKSimon commented on llvm/lib/Target/X86/X86ScheduleZnver4.td:
```diff
@@ -1341,7 +1341,7 @@ defm : Zn4WriteResXMMPair<WriteAESKeyGen, [Zn4FPAES01], 4, [1], 1>; // Key Gener
 // Carry-less multiplication instructions.
 defm : Zn4WriteResXMMPair<WriteCLMul, [Zn4FPCLM01], 4, [4], 4>;
 
-// EMMS/FEMMS
```
keep these - FEMMS still uses this sched class

RKSimon commented on llvm/test/CodeGen/X86/expand-vr64-gr64-copy.mir:
```diff
@@ -7,7 +7,7 @@
   define <2 x i32> @test_pswapdsi(<2 x i32> %a) nounwind readnone {
```
update test title

RKSimon reviewed (approved):
LGTM with a few minors

jyknight committed:
Resolve review comments.

jyknight committed:
Merge remote-tracking branch 'origin/main' into remove-mmx-3dnow

jyknight committed:
Fix formatting nit.

nathanchance commented:
I am seeing a lot of

```
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
```

while building the Linux kernel (which has `-mno-3dnow` in `arch/x86/Makefile`), is this intentional behavior? I don't have a standalone reproducer yet (I am dealing with a lot of other breakage at the moment) but I wanted to inquire on whether or not that should be happening. It appears that `-mno-3dnow` is still allowed, so we could not use our `cc-option` macro, we would have to remove it only when LLVM is at least 19.0.0.

jyknight commented:
Thanks for the report! Looks like I missed some 3dnow mentions in llvm/lib/TargetParser/X86TargetParser.cpp, so `-mno-mmx` is still emitting this feature.

I don't believe it should have any impact other than the warning spam, but I will fix it.

bnbarham commented:
`mm3dnow.h` is [included](https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/module.modulemap#L125) in the `_Builtin_intrinsics ` module, which means that anything including any intrinsic header with modules enabled will now see this warning. Any thoughts on the best way to handle this @jyknight? Some possible solutions:
1. Drop the warning entirely (seems like we've done this in the past - https://github.com/llvm/llvm-project/commit/6273b5cbcdd346a833120c55061ab56f61827068)
2. Gate the warning behind a modules check too
3. Always enable _CLANG_DISABLE_CRT_DEPRECATION_WARNINGS when modules are enabled
4. Drop the header from the modulemap
5. Move the header to its own module

CC @ian-twilightcoder and @vsapsai

jyknight commented:
I think we can just switch it to be a textual header; it no longer has any decls.

ian-twilightcoder commented:
> I think we can just switch it to be a textual header; it no longer has any decls.

It does still have _MM3DNOW_H_INCLUDED, but we could probably remove that

jyknight commented:
Do the include guard macros make it problematic to expose as a textual header? I would not have thought that to be the case.

ian-twilightcoder commented:
> Do the include guard macros make it problematic to expose as a textual header? I would not have thought that to be the case.

It means that the header guard declaration would be compiled into every pcm that includes the header. Which is maybe fine for this one, but generally isn't good.
